python/lsst/ts/standardscripts/base_close_loop.py

```diff
@@ -0,0 +1,499 @@
+# This file is part of ts_externalcripts
+#
+# Developed for the LSST Telescope and Site Systems.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["BaseCloseLoop"]
+
+import abc
+import types
+import typing
+
+import numpy as np
+import yaml
+from lsst.ts import salobj
+from lsst.ts.observatory.control.maintel.mtcs import MTCS, MTCSUsages
+from lsst.ts.observatory.control.utils.enums import ClosedLoopMode, DOFComponent
+
+STD_TIMEOUT = 10
+
+
+class BaseCloseLoop(salobj.BaseScript, metaclass=abc.ABCMeta):
+    """Closed loop script. This script is used to perform measurements of
+    the wavefront error, then propose dof offsets based on ts_ofc.
+
+    Parameters
+    ----------
+    index : `int`, optional
+        Index of Script SAL component (default=1).
+    remotes : `bool`, optional
+        Should the remotes be created (default=True)? For unit testing this
+        can be set to False, which allows one to mock the remotes behaviour.
+    descr : `str`, optional
+        Short description of the script.
+
+    Notes
+    -----
+    **Checkpoints**
+
+    - "Taking image...": If taking in-focus detection image.
+    - "[N/MAX_ITER]: Closed loop starting...": Before each closed loop
+        iteration, where "N" is the iteration number and "MAX_ITER"
+        is the maximum number of iterations.
+    - "[N/MAX_ITER]: Closed converged.": Once Closed Loop reaches convergence.
+    - "[N/MAX_ITER]: Closed applying correction.": Just before
+        corrections are applied.
+
+    **Details**
+
+    This script is used to perform measurements of the wavefront error, then
+    propose dof offsets based on ts_ofc. The offsets are not applied
+    automatically and must be turned on by the user through
+    apply_corrections attribute.
```

Comment:
Please, add a note here that if `apply_corrections` is off, the script will take a series of intra/extra focal data instead, and the number of pairs is the number of maximum iterations.